src/Middleware/LimitHandlersMiddleware.php

```diff
@@ -0,0 +1,158 @@
+ * processes one request after another without any concurrency:
+ *
+ * ```php
+ * $server = new StreamingServer(new MiddlewareRunner([
+ *     new LimitHandlersMiddleware(100), // 100 concurrent buffering handlers
+ *     new RequestBodyBufferMiddleware(2 * 1024 * 1024), // 2 MiB per request
+ *     new RequestBodyParserMiddleware(),
+ *     new LimitHandlersMiddleware(1), // only execute 1 handler (no concurrency)
+ *     $handler
+ * ]));
+ * ```
+ *
+ * @see RequestBodyBufferMiddleware
+ */
+final class LimitHandlersMiddleware
+{
+    private $limit;
+    private $pending = 0;
+    private $queue = array();
+
+    /**
+     * @param int $limit Maximum amount of concurrent requests handled.
+     *
+     * For example when $limit is set to 10, 10 requests will flow to $next
+     * while more incoming requests have to wait until one is done.
+     */
+    public function __construct($limit)
+    {
+        $this->limit = $limit;
+    }
+
+    public function __invoke(ServerRequestInterface $request, $next)
+    {
+        $body = $request->getBody();
+        if ($body instanceof ReadableStreamInterface) {
+            // pause actual body to stop emitting data until the handler is called
+            $size = $body->getSize();
+            $body = new PauseBufferStream($body);
+            $body->pauseImplicit();
+
+            // replace with buffering body to ensure any readable events will be buffered
+            $request = $request->withBody(new HttpBodyStream(
+                $body,
+                $size
+            ));
+        }
+
+        // get next queue position
+        $queue =& $this->queue;
+        $queue[] = null;
+        end($queue);
+        $id = key($queue);
+
+        $deferred = new Deferred(function ($_, $reject) use (&$queue, $id) {
+            // queued promise cancelled before its next handler is invoked
+            // remove from queue and reject explicitly
+            unset($queue[$id]);
+            $reject(new \RuntimeException('Cancelled queued next handler'));
+        });
+
+        // queue request and process queue if pending does not exceed limit
+        $queue[$id] = $deferred;
+        $this->processQueue();
+
+        $that = $this;
+        $pending = &$this->pending;
+        return $deferred->promise()->then(function () use ($request, $next, $body, &$pending) {
+            $pending++;
+
+            $ret = $next($request);
+
+            // resume readable stream and replay buffered events
+            if ($body instanceof PauseBufferStream) {
+                $body->resumeImplicit();
+            }
+
+            return $ret;
+        })->then(function ($response) use ($that, &$pending) {
+            $pending--;
+            $that->processQueue();
+
+            return $response;
+        }, function ($error) use ($that, &$pending) {
+            $pending--;
+            $that->processQueue();
+
+            throw $error;
```

Comment:
This seems to cause build errors on PHP versions below 7: https://travis-ci.org/reactphp/http/jobs/313201152#L517

Reply:
Indeed, updated to explicitly `reject()` in case the rejection value is not an `Exception`: https://github.com/reactphp/http/pull/272/files#diff-ab60bdad04437904e8a3c3e4c8eff52fR137 :shipit: